rustls/src/verify.rs

```diff
@@ -457,60 +459,113 @@ impl UnparsedCertRevocationList {
     }
 }
 
-/// A `ClientCertVerifier` that will ensure that every client provides a trusted
-/// certificate, without any name checking. Optionally, client certificates will
-/// have their revocation status checked using the DER encoded CRLs provided.
-pub struct AllowAnyAuthenticatedClient {
+/// A client certificate verifier that uses the `webpki` crate[^1] to perform client certificate
+/// validation. It must be created via the [WebpkiClientVerifier::builder()] function.
+///
+/// Once built, the provided `Arc<dyn ClientCertVerifier>` can be used with a Rustls [crate::server::ServerConfig]
+/// to configure client certificate validation using [`with_client_cert_verifier`][crate::ConfigBuilder<ClientConfig, WantsVerifier>::with_client_cert_verifier].
+///
+/// Example:
+///
+/// To require all clients present a client certificate issued by a trusted CA:
+/// ```no_run
+/// # use rustls::RootCertStore;
+/// # use rustls::server::WebpkiClientVerifier;
+/// # let roots = RootCertStore::empty();
+/// let client_verifier = WebpkiClientVerifier::builder()
+///   .with_roots(roots)
+///   .build();
+/// ```
+///
+/// Or, to allow clients presenting a client certificate authenticated by a trusted CA, or
+/// anonymous clients that present no client certificate:
+/// ```no_run
+/// # use rustls::RootCertStore;
+/// # use rustls::server::WebpkiClientVerifier;
+/// # let roots = RootCertStore::empty();
+/// let client_verifier = WebpkiClientVerifier::builder()
+///   .with_roots(roots)
+///   .allow_unauthenticated()
+///   .build();
+/// ```
+///
+/// If you wish to perform **no** client authentication at all:
+/// ```no_run
+/// # use rustls::RootCertStore;
+/// # use rustls::server::WebpkiClientVerifier;
+/// # let roots = RootCertStore::empty();
+/// let client_verifier = WebpkiClientVerifier::builder()
+///   .allow_unauthenticated()
+///   .build();
+/// ```
+///
+/// You can also configure the client verifier to check for certificate revocation with
+/// client certificate revocation lists (CRLs):
+/// ```no_run
+/// # use rustls::RootCertStore;
+/// # use rustls::server::{WebpkiClientVerifier};
+/// # let roots = RootCertStore::empty();
+/// # let crls = Vec::new();
+/// let client_verifier = WebpkiClientVerifier::builder()
+///   .with_roots(roots)
+///   .with_crls(crls)
+///   .expect("invalid crls")
+///   .build();
+/// ```
+///
+/// [^1]: <https://github.com/rustls/webpki>
+pub struct WebpkiClientVerifier {
     roots: RootCertStore,
     subjects: Vec<DistinguishedName>,
     crls: Vec<webpki::OwnedCertRevocationList>,
+    anon_policy: AnonymousClientsPolicy,
 }
 
-impl AllowAnyAuthenticatedClient {
-    /// Construct a new `AllowAnyAuthenticatedClient`.
+impl WebpkiClientVerifier {
+    /// Create builder to build up the `webpki` client certificate verifier configuration.
+    ///
+    /// For more information, see the [`WebpkiClientVerifier`] documentation.
+    pub fn builder() -> ClientCertVerifierBuilder {
```

Comment:
> I think we should also offer a no_client_auth() constructor that short-circuits the builder here, constructing the obvious WebPkiClientVerifier without the requirement to go through builder().build(), especially since it seems like this will be the most common scenario by some distance.

Good idea, I've added this method and updated some comments throughout to point to it.

> After considering the subtle differences resulting from empty vs non-empty roots, maybe we should just enforce non-empty roots for the builder().build() path (perhaps by having builder() take a roots argument)?

I was hesitant about this at first but I implemented it and agree it makes things clearer in conjunction with the `no_client_auth` method. It's nice that this moves the chance of error to the initial construction of the builder, and when providing CRLs, and not at the time of calling `build`. I think that will be a better experience.